src/python/TaskWorker/Actions/DBSDataDiscovery.py

```diff
@@ -91,15 +92,124 @@ def requestTapeRecall(self, blockList=[], system='Dynamo', msgHead=''):   # pyli
+            # prepare container /TapeRecall/taskname/USER in the service scope
+            myScope = 'user.%s' % self.config.Services.Rucio_account
+            containerName = '/TapeRecall/%s/USER' % self.taskName.replace(':', '.')
+            containerDid = {'scope':myScope, 'name':containerName}
+            self.logger.info("Create RUcio container %s", containerName)
+            try:
+                self.rucioClient.add_container(myScope, containerName)
+            except DataIdentifierAlreadyExists:
+                self.logger.debug("Container name already exists in Rucio. Keep going")
+            except Exception as ex:
+                msg += "Rucio exception creating container: %s" %  (str(ex))
+                raise TaskWorkerException(msg)
+            try:
+                self.rucioClient.attach_dids(myScope, containerName, dids)
+            except DuplicateContent:
+                self.logger.debug("Some dids are already in this container. Keep going")
+            except Exception as ex:
+                msg += "Rucio exception adding blocks to container: %s" %  (str(ex))
+                raise TaskWorkerException(msg)
+            self.logger.info("Rucio container %s:%s created with %d blocks", myScope, containerName, len(blockList))
+
+            # Compute size of recall request
+            sizeToRecall = 0
+            for block in blockList:
+                replicas = self.rucioClient.list_dataset_replicas('cms', block)
+                blockBytes = replicas.next()['bytes']  # pick first replica for each block, they better all have same size
+                sizeToRecall += blockBytes
+            TBtoRecall = sizeToRecall // 1e12
+            if TBtoRecall > 0:
+                self.logger.info("Total size of data to recall : %d TBytes", TBtoRecall)
+            else:
+                self.logger.info("Total size of data to recall : %d GBytes", sizeToRecall/1e9)
+
+            if TBtoRecall > 30.:
+                grouping = 'DATASET'  # Rucio DATASET i.e. CMS block !
+                self.logger.info("Will scatter blocks on multiple sites")
+            else:
+                grouping = 'ALL'
+                self.logger.info("Will place all blocks at a single site")
+
+            # create rule
+            RSE_EXPRESSION = 'ddm_quota>0&tier=2&rse_type=DISK'
+            RSE_EXPRESSION = 'T3_IT_Trieste' # for testing
+            WEIGHT = 'ddm_quota'
+            WEIGHT = None # for testing
+            DAYS = 14 * 24 * 3600
+            ASK_APPROVAL = False
+            ASK_APPROVAL = True # for testing
+            copies = 1
+            try:
+                ruleId = self.rucioClient.add_replication_rule(dids=[containerDid],
+                                                  copies=copies, rse_expression=RSE_EXPRESSION,
+                                                  grouping=grouping,
+                                                  weight=WEIGHT, lifetime=DAYS, account=self.username,
```

Comment:
The rule is made with the CRAB user's account? Shouldn't it be the crab_tape_recall account?

Reply:
I am using an account for which I can authenticate so that I can test the code. my username is perfect for this. I am aware of the need to change but I forgot to highlight this with a comment.
Maybe we can  use existing crab_server account  ?

Reply:
Hmm how do we have two crab accounts already?
```
$ rucio-admin account info crab_tape_recall
status     : ACTIVE
account    : crab_tape_recall
account_type : SERVICE
created_at : 2020-10-15T20:57:44
updated_at : 2020-10-15T20:57:44
suspended_at : None
deleted_at : None
email      : stefano.belforte@cern.ch
$ rucio-admin account info crab_server
status     : ACTIVE
account    : crab_server
account_type : SERVICE
created_at : 2020-04-29T15:40:28
updated_at : 2020-04-29T15:40:28
suspended_at : None
deleted_at : None
email      : Stefano.Belforte@cern.ch
```
The former is used by @ericvaandering's script that the operators are using to stage data. For continuity it would be nice to keep using that one.

Reply:
I guess that when Eric wrote his script he forgot that he had created the other account already. But if you and him agree that one account to browse the data and one to create tape recall are better, fine with me. Let me know when this is final. I tried to authenticate as crab_tape_recall at some time but failed both my DN and with the CRAB service DN's.

Reply:
hmmm... @nsmith- @ericvaandering what is the relation between `account` argument inside the call to `add_replication_rule` and the one which I used earlier to create the rucio Client object ? Can they be different/independent ?

Reply:
current situation, crab_tape_recall seems something that only accepts login via password
```
belforte@lxplus797/~> rucio-admin account list-identities crab_server
Identity: /DC=ch/DC=cern/OU=computers/CN=tw/crab-prod-tw02.cern.ch,	type: X509
Identity: /DC=ch/DC=cern/OU=computers/CN=tw/crab-preprod-tw02.cern.ch,	type: X509
Identity: /DC=ch/DC=cern/OU=computers/CN=tw/crab-dev-tw01.cern.ch,	type: X509
Identity: /DC=ch/DC=cern/OU=computers/CN=tw/crab-prod-tw01.cern.ch,	type: X509
Identity: /DC=ch/DC=cern/OU=computers/CN=tw/crab-preprod-tw01.cern.ch,	type: X509
Identity: /DC=ch/DC=cern/OU=computers/CN=tw/crab-preprod-tw03.cern.ch,	type: X509
belforte@lxplus797/~> rucio-admin account list-identities crab_tape_recall
belforte@lxplus797/~> 


```

Reply:
Accounts with enough privilege can create rules on behalf of other users. Not sure if crab_server could do this on behalf of crab_tape_recall (which was created as a placeholder for this service and so that we can attach a quota if needed). 

You may need to make a second instance of the rucio Client() object. And if I am missing a DN, please let me know. Easy to add.

Reply:
maybe start by adding those identities for crab_server account to the crab_tape_recall as well ?

Reply:
Note that so far I have only tested this stuff with my personal credential in my VM. I am trying to nail all things so that I can deploy this on preproducion crab where we can test in realistic conditions.

Reply:
Your personal credential is not going to work. But can you try with a Client() instantiated with the crab_server account but creating the rule under crab_tape_recall? I think that's going to work because, as far as I know, transfer Ops are running my script under their account, not under root.

Reply:
I can surely try (and was planning to do as soon as I get this thing on preprod). Will let you know.